File: apps/ios/Ghostpass/Services/SsoMobile.swift
import AuthenticationServices
import CryptoKit
import Foundation

/// Ouvrir une session par SSO depuis l'application.
///
/// Le contrat est décrit dans `docs/sso-mobile.md`. Ce qu'il faut retenir ici :
///
/// **Le PKCE couvre le saut application ↔ serveur**, et non application ↔ fournisseur
/// d'identité. C'est délibéré : sur iOS, n'importe quelle application peut revendiquer un
/// schéma d'URL personnalisé, donc le retour destiné à GhostPass est interceptable. Le code
/// qui y transite ne vaut rien sans le vérificateur, que seule cette application détient.
///
/// Le SSO **authentifie ; il n'ouvre pas le coffre.** Celui-ci reste scellé sous le mot de
/// passe maître, qui sera demandé ensuite. Les confondre est la première erreur de
/// conception d'un client à connaissance nulle.
enum SsoMobile {
    /// Le schéma que l'`ASWebAuthenticationSession` attend en retour.
    ///
    /// Il n'est **pas** déclaré dans `Info.plist`, et c'est voulu : le déclarer ferait de
    /// GhostPass un gestionnaire général de ce schéma, alors que seule la session
    /// d'authentification en cours doit le recevoir.
    static let schema = "ch.stackops.ghostpass"
    static let adresseDeRetour = "\(schema)://sso"

    // ─── PKCE ───

    /// Un vérificateur et son défi, liés par SHA-256.
    struct Pkce {
        let verificateur: String
        let defi: String

        /// Tire un vérificateur aléatoire et calcule son défi.
        init() {
            self.init(verificateur: Self.chaineAleatoire(octets: 32))
        }

        /// Utilisable avec un vérificateur imposé, pour éprouver le calcul contre le
        /// vecteur de la RFC 7636.
        init(verificateur: String) {
            self.verificateur = verificateur
            self.defi = Self.defi(pour: verificateur)
        }

        /// `base64url(sha256(vérificateur))`, sans remplissage.
        ///
        /// SHA-256 rend 32 octets, soit exactement 43 caractères en base64url — la
        /// longueur que le serveur exige. Laisser le remplissage `=` en ferait 44 et le
        /// `start` refuserait, ce qui est le bon comportement : un défi mal formé doit
        /// échouer là où il arrive, pas trois étapes plus loin.
        static func defi(pour verificateur: String) -> String {
            let empreinte = SHA256.hash(data: Data(verificateur.utf8))
            return base64url(Data(empreinte))
        }

        static func chaineAleatoire(octets: Int) -> String {
            var brut = Data(count: octets)
            let resultat = brut.withUnsafeMutableBytes {
                SecRandomCopyBytes(kSecRandomDefault, octets, $0.baseAddress!)
            }
            // `SecRandomCopyBytes` ne peut échouer qu'en cas de défaillance du système ;
            // retomber sur un aléa faible serait pire que s'arrêter, puisque tout le
            // dispositif repose sur l'imprévisibilité de cette valeur.
            precondition(resultat == errSecSuccess, "Le générateur aléatoire du système a échoué.")
            return base64url(brut)
        }

        static func base64url(_ donnees: Data) -> String {
            donnees.base64EncodedString()
                .replacingOccurrences(of: "+", with: "-")
                .replacingOccurrences(of: "/", with: "_")
                .replacingOccurrences(of: "=", with: "")
        }
    }

    // ─── Le retour, et sa lecture ───

    enum ErreurDeRetour: LocalizedError, Equatable {
        /// Le retour ne porte pas de code. C'est **l'unique** critère d'échec.
        case sansCode(motif: String?)
        /// L'état rendu n'est pas celui qu'on a envoyé : ce retour ne répond pas à notre
        /// demande, et l'accepter reviendrait à ouvrir une session qu'un tiers a lancée.
        case etatInattendu

        var errorDescription: String? {
            switch self {
            case .etatInattendu:
                return String(localized: "La réponse ne correspond pas à la demande. Réessayez.")
            case .sansCode(let motif):
                switch motif {
                case "not_provisioned":
                    // Le serveur ne crée jamais de compte par SSO : le coffre est scellé
                    // sous le mot de passe maître, donc un compte provisionné à la volée
                    // n'aurait rien à ouvrir.
                    return String(localized: "Cette adresse n'a pas de compte GhostPass sur ce serveur.")
                default:
                    return String(localized: "L'authentification a échoué.")
                }
            }
        }
    }

    /// Lit le retour du navigateur.
    ///
    /// **L'échec est l'absence de `code`, pas la présence d'`error`.** Un client qui teste
    /// `error` et poursuit sinon appellerait l'échange avec un code vide, et lirait le refus
    /// du serveur comme une panne réseau plutôt que comme un rejet d'authentification.
    static func codeDuRetour(_ url: URL, etatAttendu: String) -> Result<String, ErreurDeRetour> {
        let elements = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
        func valeur(_ nom: String) -> String? {
            let brut = elements.first { $0.name == nom }?.value
            return (brut?.isEmpty ?? true) ? nil : brut
        }

        // L'état se vérifie avant tout le reste : un retour qui n'est pas le nôtre ne
        // mérite pas qu'on lise ce qu'il transporte.
        guard valeur("state") == etatAttendu else { return .failure(.etatInattendu) }
        guard let code = valeur("code") else { return .failure(.sansCode(motif: valeur("error"))) }
        return .success(code)
    }

    /// L'adresse d'ouverture du flux, telle que le serveur l'attend.
    static func adresseDeDepart(serveur: URL, defi: String, etat: String) -> URL? {
        var composants = URLComponents(
            url: serveur.appendingPathComponent("api/auth/sso/mobile/start"),
            resolvingAgainstBaseURL: false)
        composants?.queryItems = [
            URLQueryItem(name: "code_challenge", value: defi),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: etat),
            URLQueryItem(name: "redirect_uri", value: adresseDeRetour),
        ]
        return composants?.url
    }

    // ─── Le navigateur ───

    /// Ouvre la session web et rend le code, ou jette.
    @MainActor
    static func obtenirUnCode(serveur: URL, pkce: Pkce, ancre: ASPresentationAnchor) async throws
        -> String
    {
        let etat = Pkce.chaineAleatoire(octets: 16)
        guard let depart = adresseDeDepart(serveur: serveur, defi: pkce.defi, etat: etat) else {
            throw APIError.badURL
        }

        let retour: URL = try await withCheckedThrowingContinuation { suite in
            let session = ASWebAuthenticationSession(
                url: depart, callbackURLScheme: schema
            ) { url, erreur in
                if let url {
                    suite.resume(returning: url)
                } else {
                    suite.resume(throwing: erreur ?? ErreurDeRetour.sansCode(motif: nil))
                }
            }
            session.presentationContextProvider = Ancre.partagee(ancre)
            // Session éphémère : les cookies du fournisseur d'identité ne sont pas partagés
            // avec Safari. Sur un téléphone prêté, une session laissée ouverte permettrait à
            // la personne suivante d'entrer sans rien saisir.
            session.prefersEphemeralWebBrowserSession = true
            session.start()
        }

        switch codeDuRetour(retour, etatAttendu: etat) {
        case .success(let code): return code
        case .failure(let erreur): throw erreur
        }
    }

    /// `ASWebAuthenticationSession` exige un fournisseur d'ancre retenu en mémoire : passé
    /// en local, il serait libéré avant que la session s'affiche.
    private final class Ancre: NSObject, ASWebAuthenticationPresentationContextProviding {
        private static var vivante: Ancre?
        private let fenetre: ASPresentationAnchor

        private init(_ fenetre: ASPresentationAnchor) { self.fenetre = fenetre }

        static func partagee(_ fenetre: ASPresentationAnchor) -> Ancre {
            let ancre = Ancre(fenetre)
            vivante = ancre
            return ancre
        }

        func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
            fenetre
        }
    }
}

File: apps/android/noyau/src/main/kotlin/ch/stackops/ghostpass/SsoMobile.kt
package ch.stackops.ghostpass

import java.security.MessageDigest
import java.security.SecureRandom

/**
 * Ouvrir une session par SSO depuis l'application.
 *
 * Le contrat est `docs/sso-mobile.md` (branche `main` du serveur), et le pendant iOS est
 * `apps/ios/Ghostpass/Services/SsoMobile.swift`. Ce qu'il faut retenir ici :
 *
 * **Le PKCE couvre le saut application ↔ serveur**, et non application ↔ fournisseur
 * d'identité. C'est délibéré, et pour deux raisons qui se cumulent :
 *
 *  - l'IdP est Cloudflare Access for SaaS, qui n'enregistre aucun client OIDC public :
 *    chaque application reçoit un `client_secret`, donc l'application ne peut pas mener
 *    l'échange elle-même ;
 *  - sur Android comme sur iOS, **n'importe quelle application peut revendiquer un schéma
 *    d'URL personnalisé**. Le retour destiné à GhostPass est donc interceptable. S'il
 *    portait un jeton de session, l'intercepteur aurait le compte ; il ne porte qu'un code,
 *    qui ne vaut rien sans le vérificateur que seule cette application détient.
 *
 * Le SSO **authentifie ; il n'ouvre pas le coffre**. Celui-ci reste scellé sous le mot de
 * passe maître, demandé ensuite. Les confondre est la première erreur de conception d'un
 * client à connaissance nulle.
 *
 * ## Ce qui n'est pas de la cryptographie réécrite
 *
 * SHA-256 et le tirage aléatoire viennent de la plateforme (`MessageDigest`,
 * `SecureRandom`), comme ils viennent de CryptoKit côté iOS. Ce n'est pas une
 * implémentation du produit : c'est un condensé normalisé par la RFC 7636, dont le vecteur
 * de référence est dans [SsoMobileTest]. Le cœur Rust n'expose pas SHA-256 nu, et le lui
 * faire exposer pour ce seul usage coûterait plus qu'il ne protège.
 */
object SsoMobile {

    /**
     * Le schéma de retour, tel que le serveur l'a en liste blanche
     * (`SSO_MOBILE_REDIRECT_URIS`, défaut `ch.stackops.ghostpass://sso`).
     *
     * §8 du brief préfère un App Link vérifié (`https://`) quand l'instance le permet : lui
     * seul est réellement exclusif. Ce n'est **pas** ce qui est fait ici, et c'est un écart
     * assumé — la liste blanche du serveur est réglée sur ce schéma, et un App Link
     * exigerait de publier un `assetlinks.json` sur chaque instance auto-hébergée. C'est
     * précisément le genre d'exigence qu'un produit auto-hébergeable ne peut pas poser.
     *
     * Ce qui rend l'écart acceptable est le PKCE : un code intercepté est inerte.
     */
    const val SCHEMA = "ch.stackops.ghostpass"
    const val ADRESSE_DE_RETOUR = "$SCHEMA://sso"

    // ─── PKCE ───

    /** Un vérificateur et son défi, liés par SHA-256. */
    class Pkce private constructor(val verificateur: String) {
        /** `base64url(sha256(vérificateur))`, sans remplissage. */
        val defi: String = defiPour(verificateur)

        companion object {
            /** Tire un vérificateur aléatoire de 32 octets. */
            fun tirer(): Pkce = Pkce(chaineAleatoire(32))

            /** Utilisable avec un vérificateur imposé, pour éprouver contre la RFC 7636. */
            fun avec(verificateur: String): Pkce = Pkce(verificateur)
        }
    }

    /**
     * `base64url(sha256(v))` sans remplissage.
     *
     * SHA-256 rend 32 octets, soit exactement **43 caractères** en base64url — la longueur
     * que le serveur exige. Laisser le remplissage `=` en ferait 44 et le `start`
     * refuserait, ce qui est le bon comportement : un défi mal formé doit échouer là où il
     * arrive, pas trois étapes plus loin.
     */
    fun defiPour(verificateur: String): String =
        base64url(MessageDigest.getInstance("SHA-256").digest(verificateur.toByteArray(Charsets.UTF_8)))

    /**
     * Une chaîne aléatoire en base64url.
     *
     * `SecureRandom` et rien d'autre : tout le dispositif repose sur l'imprévisibilité de
     * cette valeur, et retomber sur un aléa faible serait pire que s'arrêter.
     */
    fun chaineAleatoire(octets: Int): String =
        base64url(ByteArray(octets).also { SecureRandom().nextBytes(it) })

    private fun base64url(donnees: ByteArray): String =
        java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(donnees)

    // ─── L'ouverture du flux ───

    /**
     * L'adresse d'ouverture, telle que le serveur l'attend.
     *
     * Le `state` est **opaque et tiré par l'application** : le serveur le range tel quel
     * pendant dix minutes dans `auth_ephemeral`. Y glisser une adresse de courriel ou un
     * identifiant de compte y écrirait une donnée personnelle dans un réceptacle qui n'est
     * pas prévu pour en porter — le serveur ne peut pas l'en empêcher.
     */
    fun adresseDeDepart(serveur: String, defi: String, etat: String): String {
        val base = serveur.removeSuffix("/")
        return base + "/api/auth/sso/mobile/start" +
            "?code_challenge=" + encoder(defi) +
            "&code_challenge_method=S256" +
            "&state=" + encoder(etat) +
            "&redirect_uri=" + encoder(ADRESSE_DE_RETOUR)
    }

    // ─── Le retour, et sa lecture ───

    /** Ce qui peut mal tourner au retour du navigateur. */
    sealed interface EchecDeRetour {
        /**
         * L'état rendu n'est pas celui qu'on a envoyé.
         *
         * Ce retour ne répond pas à *notre* demande, et l'accepter reviendrait à ouvrir une
         * session qu'un tiers a lancée. **Vérifié en premier** : un retour qui n'est pas le
         * nôtre ne mérite pas qu'on lise ce qu'il transporte.
         */
        data object EtatInattendu : EchecDeRetour

        /** Le retour ne porte pas de code. C'est **l'unique** critère d'échec restant. */
        data class SansCode(val motif: String?) : EchecDeRetour
    }

    /**
     * Lit le retour du navigateur.
     *
     * **L'échec est l'absence de `code`, pas la présence d'`error`.** Un client qui teste
     * `error` et poursuit sinon appellerait l'échange avec un code vide, et lirait le refus
     * du serveur comme une panne réseau plutôt que comme un rejet d'authentification. Le
     * serveur, lui, renvoie toujours par le schéma de l'application — une page d'erreur y
     * laisserait l'utilisateur bloqué dans une fenêtre qui ne se referme pas.
     */
    fun codeDuRetour(url: String, etatAttendu: String): Result<String> {
        val parametres = parametres(url)
        fun valeur(nom: String): String? = parametres[nom]?.ifEmpty { null }

        if (valeur("state") != etatAttendu) {
            return Result.failure(ErreurDeRetour(EchecDeRetour.EtatInattendu))
        }
        val code = valeur("code")
            ?: return Result.failure(ErreurDeRetour(EchecDeRetour.SansCode(valeur("error"))))
        return Result.success(code)
    }

    /** Une exception porteuse d'un [EchecDeRetour], pour tenir dans un `Result`. */
    class ErreurDeRetour(val echec: EchecDeRetour) : Exception(message(echec)) {
        companion object {
            fun message(echec: EchecDeRetour): String = when (echec) {
                EchecDeRetour.EtatInattendu ->
                    "La réponse ne correspond pas à la demande. Réessayez."
                is EchecDeRetour.SansCode -> when (echec.motif) {
                    // Le serveur ne crée jamais de compte par SSO : le coffre est scellé
                    // sous le mot de passe maître, donc un compte provisionné à la volée
                    // n'aurait rien à ouvrir — l'utilisateur verrait un coffre vide et
                    // croirait avoir perdu ses données.
                    "not_provisioned" ->
                        "Cette adresse n'a pas de compte GhostPass sur ce serveur."
                    else -> "L'authentification a échoué."
                }
            }
        }
    }

    /**
     * Les paramètres de requête d'une URL de retour, déséchappés.
     *
     * Écrit à la main pour la même raison que [LienOtpauth] : ce qui arrive ici vient du
     * dehors, et `java.net.URI` lève sur des formes qu'il vaut mieux refuser proprement.
     */
    private fun parametres(url: String): Map<String, String> {
        val requete = url.substringAfter('?', "").substringBefore('#')
        if (requete.isEmpty()) return emptyMap()
        val resultat = LinkedHashMap<String, String>()
        for (couple in requete.split('&')) {
            val egal = couple.indexOf('=')
            if (egal < 0) continue
            resultat[decoder(couple.substring(0, egal))] = decoder(couple.substring(egal + 1))
        }
        return resultat
    }

    private fun encoder(valeur: String): String =
        java.net.URLEncoder.encode(valeur, "UTF-8").replace("+", "%20")

    private fun decoder(valeur: String): String =
        try {
            java.net.URLDecoder.decode(valeur, "UTF-8")
        } catch (_: Exception) {
            valeur
        }
}
